File: Assets/Scripts/FlyDropMovement.fs
namespace MegaDark
open UnityEngine

type FlyDropMovement() =
    inherit MonoBehaviour()

    [<SerializeField>] 
    let mutable flySpeed = -2.0f
    [<SerializeField>]
    let mutable dropInitialSpeed = 2.0f
    [<SerializeField>]
    let mutable dropAcceleration = -5.0f

    let mutable physics : SpritePhysics = null

    let mutable isFalling = false

    let shouldFall (delta : Vector3) = delta.x * flySpeed < 0.0f

    let SetVel x y =
        physics.vel <- Vector2(x, y)

    let FlyUpdate pos =
        let player = PlayerManager.Instance.Player
        let playerPos = player.transform.position
        let delta = playerPos - pos
        SetVel flySpeed 0.0f

        isFalling <- shouldFall delta

    let FallUpdate() =
        SetVel 0.0f dropInitialSpeed

    member this.Start() =
        physics <- this.GetComponent<SpritePhysics>()

    member this.FixedUpdate() =
        if not isFalling
        then FlyUpdate this.transform.position
        else FallUpdate()


type BottomlessPit() =
    inherit MonoBehaviour()

    interface Collideable with
        member this.OnCollide collision =
            let health = collision.sender.GetComponent<Health>()
            health.Kill()

File: Assets/Scripts/Enemy/FlyDropMovement.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class FlyDropMovement : MonoBehaviour {
	[SerializeField] private float flySpeed = -2.0f;
    [SerializeField] private float dropInitialSpeed = 2.0f;
    [SerializeField] private float dropAcceleration = -5.0f;

    private SpritePhysics physics = null;

    private bool isFalling = false;

    void Start() {
        physics = GetComponent<SpritePhysics>();
	}

    void FixedUpdate() {
        if (!isFalling) {
	        flyUpdate();
		}
        else {
			fallUpdate();
		}
	}

    private void flyUpdate() {
        setVel(flySpeed, 0.0f);

		possiblyStartFalling();
	}

	private void possiblyStartFalling() {
		if (shouldFall()) {
			isFalling = true;
			setVel(0.0f, dropInitialSpeed);
		}
	}

    private bool shouldFall() {
        GameObject player = PlayerManager.Instance.Player;
        SpritePhysics playerPhysics = player.GetComponent<SpritePhysics>();
        Vector3 playerPos = player.transform.position;
        Vector3 delta = playerPos - transform.position;

        float t = timeToFallFromHeight(delta.y);

        return (delta.x + t * playerPhysics.Vel.x) * flySpeed < 0.0f;
    }

    private float timeToFallFromHeight(float height) {
        // Given h = v_0 * t + 1/2 * a * t^2 ,
        // solve quadratic formula to find t = (-v_0 - sqrt(v_0^2 + 2*a*h)) / a
        float v0 = dropInitialSpeed;
        float a = dropAcceleration;
        return (-v0 - Mathf.Sqrt(v0 * v0 + 2 * a * height)) / a;
    }

    private void setVel(float x, float y) {
        physics.Vel = new Vector2(x, y);
	}

    private void fallUpdate() {
		float vel = physics.Vel.y;
		vel += Time.fixedDeltaTime * dropAcceleration;
        setVel(0.0f, vel);
	}
}
